crates/sui-types/src/bls12381.rs

```diff
@@ -0,0 +1,742 @@
+use serde_with::Bytes;
+use sha3::Sha3_256;
+use std::borrow::Borrow;
+use std::collections::{HashMap, HashSet};
+use std::hash::{Hash, Hasher};
+use std::slice;
+use std::str::FromStr;
+use std::{
+    fmt::{self, Display},
+    mem::MaybeUninit,
+};
+use hkdf::Hkdf;
+
+use blst::{min_pk as bls, BLST_ERROR};
+use ::blst::{blst_scalar, blst_scalar_from_uint64};
+use rand::{RngCore};
+
+pub const BLST_SK_SIZE: usize = 32;
+pub const BLST_PK_SIZE: usize = 48;
+pub const BLST_SIG_SIZE: usize = 96;
+
+pub type Bls12381PublicKey = bls::PublicKey;
+
+#[derive(Debug)]
+pub struct BlstKeypair {
+    pub public: bls::PublicKey,
+    pub secret: bls::SecretKey 
+}
+
+impl BlstKeypair {
+    fn to_bytes(&self) -> Vec<u8> {
+        let key_concat: Vec<u8> = self.public.to_bytes()
+        .iter().cloned()
+        .chain(self.secret.to_bytes().iter().cloned())
+        .collect();
+        key_concat
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<BlstKeypair, BLST_ERROR> {
+        let public = bls::PublicKey::from_bytes(&bytes[..BLST_PK_SIZE])?;
+        let secret = bls::SecretKey::from_bytes(&bytes[BLST_PK_SIZE..])?;
+        Ok(BlstKeypair {
+            public,
+            secret
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Bls12381KeyPair {
+    pub key_pair: BlstKeypair,
+    pub public_key_cell: OnceCell<Bls12381PublicKeyBytes>,
+}
+
+impl Bls12381KeyPair {
+    pub fn public_key_bytes(&self) -> &Bls12381PublicKeyBytes {
+        self.public_key_cell.get_or_init(|| {
+            let pk_arr: [u8; BLST_PK_SIZE] = self.key_pair.public.to_bytes();
+            Bls12381PublicKeyBytes(pk_arr)
+        })
+    }
+
+    #[must_use]
+    pub fn copy(&self) -> Self {
+        Bls12381KeyPair {
+            key_pair: BlstKeypair {
+                public: bls::PublicKey::from_bytes(&self.key_pair.public.to_bytes()).unwrap(),
+                secret: bls::SecretKey::from_bytes(&self.key_pair.secret.to_bytes()).unwrap(),
+            },
+            public_key_cell: OnceCell::new(),
+        }
+    }
+
+    pub fn new_deterministic_keypair(seed: &[u8], id: &[u8], domain: &[u8]) -> SuiResult<Bls12381KeyPair>{
+        // HKDF<Sha3_256> to deterministically generate an ed25519 private key.
+        let hk = Hkdf::<Sha3_256>::new(Some(id), seed);
+        let mut okm = [0u8; BLST_SK_SIZE];
+        hk.expand(domain, &mut okm)
+            .map_err(|e| SuiError::HkdfError(e.to_string()))?;
+
+        let secret = bls::SecretKey::key_gen(&okm, &[])
+            .map_err(|e| SuiError::SignatureKeyGenError("Error Filler".to_string()))?;
+        let public = secret.sk_to_pk();
+
+        Ok(Bls12381KeyPair {
+            key_pair: BlstKeypair {
+                public,
+                secret 
+            },
+            public_key_cell: OnceCell::new()
+        })
+    }
+
+    /// Make a Narwhal-compatible key pair from a Sui keypair.
+    pub fn make_narwhal_keypair(&self) -> Ed25519KeyPair {
+        let kp = DalekKeypair::generate(&mut OsRng);
```

Comment:
We generate a random Ed25519 KP for now, until we also refactor the Narwhal repo so that it supports BLS (@huitseeker Would ask you about how easy this would be)

Reply:
The Narwhal repo has been supporting BLS12-381 since January.